uxarray/io/_geopandas.py

```diff
@@ -0,0 +1,267 @@
+
+    grid_ds["node_lon"] = xr.DataArray(
+        data=node_lon, dims=ugrid.NODE_DIM, attrs=ugrid.NODE_LON_ATTRS
+    )
+    grid_ds["node_lat"] = xr.DataArray(
+        data=node_lat, dims=ugrid.NODE_DIM, attrs=ugrid.NODE_LAT_ATTRS
+    )
+    grid_ds["face_node_connectivity"] = xr.DataArray(
+        data=connectivity,
+        dims=ugrid.FACE_NODE_CONNECTIVITY_DIMS,
+        attrs=ugrid.FACE_NODE_CONNECTIVITY_ATTRS,
+    )
+
+    return grid_ds, None
+
+
+def _gpd_read(filepath, driver=None, **kwargs):
+    """Read a geospatial data using geopandas.
+
+    Parameters
+    ----------
+    filepath : str
+        Filepath to geospatial data.
+    driver : str, optional
+        Driver to use, by default None, in which case geopandas will try to infer the driver from the file extension.
+    **kwargs
+        Keyword arguments to pass to geopandas.read_file().
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        GeoDataFrame with geometries.
+    int
+        Maximum number of nodes in a polygon/multipolygon.
+    """
+    try:
+        gdf = gpd.read_file(filepath, driver=driver, **kwargs)
+        gdf = _set_crs(gdf)
+    except Exception as e:
+        print(f"An error occurred while reading the geospatial data: {e}")
+
+    max_polygon_nodes = gdf["geometry"].apply(_get_num_nodes).max()
+
+    return gdf, max_polygon_nodes
+
+
+def _set_crs(gdf):
+    """Set CRS for GeoDataFrame if not already set.
+
+    Parameters
+    ----------
+    gdf : gpd.GeoDataFrame
+        GeoDataFrame to set CRS for.
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        GeoDataFrame with CRS set.
+    """
+    if gdf.crs is None:
+        gdf = gdf.set_crs("EPSG:4326")
+        print("Original CRS: None\nAssigned CRS:", gdf.crs)
+
+    if gdf.crs != "EPSG:4326":
+        gdf = gdf.to_crs("EPSG:4326")
+        print("Transformed CRS:", gdf.crs)
+
+    return gdf
+
+
+def _extract_geometry_info(gdf, max_coord_size):
+    """Extract node and connectivity information from GeoDataFrame.
+
+    Parameters
+    ----------
+    gdf : gpd.GeoDataFrame
+        GeoDataFrame with geometries.
+    max_coord_size : int
+        Maximum number of nodes in a polygon/multipolygon.
+
+    Returns
+    -------
+    np.ndarray
+        Array of longitudes.
+    np.ndarray
+        Array of latitudes.
+    np.ndarray
+        Connectivity array.
+    """
+    node_lon = np.array([])
+    node_lat = np.array([])
+    connectivity = np.empty((0, max_coord_size - 1), dtype=INT_DTYPE)
+
+    node_index = 0
+
+    for _, row in gdf.iterrows():
+        geometry = row["geometry"]
+        if geometry.geom_type == "Polygon":
+            node_lat, node_lon, connectivity, node_index = _read_polygon(
+                geometry, node_lat, node_lon, connectivity, node_index
+            )
+        elif geometry.geom_type == "MultiPolygon":
+            node_lat, node_lon, connectivity, node_index = _read_multipolygon(
```

Comment:
sometimes they are very high 100's as geojson and shapefiles have 2D shapes covering boundaries.

Reply:
Then, maybe a few ideas to optimize this:

- If you know the end size of the numpy arrays, node_lon and node_lat, create a numpy array in that size upfront, then perform append 
- Otherwise, append to a list rather than numpy array. At the very end of the operation, convert list to numpy array

What do you think?

Reply:
> Then, maybe a few ideas to optimize this:
> 
> * If you know the end size of the numpy arrays, node_lon and node_lat, create a numpy array in that size upfront, then perform append
> * Otherwise, append to a list rather than numpy array. At the very end of the operation, convert list to numpy array
> 
> What do you think?

Using lists as  list is an amortized O(1), thanks for this suggestion I have just pushed changes.